#' Read meth, UNmeth, and meth_level data, the corresponding column name is \\.name$
#'
#' @param file_tmp DNA methylation or chromatin accessibility matrix
#' @param string_data Sample type information, such as meth, UNmeth, methlevel
#'
#' @return Sample data after selecting columns
#' @export
#'
#' @examples
Read_file_meth_colname <- function(file_tmp,string_data){
  file_data <- file_tmp
  string_tmp <- string_data
  string_col <- paste0("\\.",string_tmp)
  string_col2 <- paste0(string_col,"$")

  column_indices <- grep(string_col2, names(file_data), value = TRUE)
  file_data_meth <- file_data[,column_indices]
  # file_data_meth <- cbind(rownames(file_data),file_data_meth)
  rownames(file_data_meth) <- rownames(file_tmp)
  return(file_data_meth)
}


#' MOFA object is created
#'
#' @param file_tmp DNA methylation or chromatin accessibility level data
#' @param group Sample grouping data
#'
#' @return a trained MOFA object
#' @export
#'
#' @examples
MOFA_est <- function(file_tmp,group=NULL){
  mofa_list <- file_tmp


  if (is.null(group)) {
    MOFAObject <- create_mofa(mofa_list)
  }else if(!is.null(group)){
    MOFAObject <- create_mofa(mofa_list,group=group)
  }

  ModelOptions <- get_default_model_options(MOFAObject)
  TrainOptions <- get_default_training_options(MOFAObject)
  DataOptions <- get_default_data_options(MOFAObject)

  MOFAObject <- prepare_mofa(MOFAObject,
                             model_options = ModelOptions,
                             training_options = TrainOptions,
                             data_options = DataOptions)

  MOFAObject <- run_mofa(MOFAObject)
  return(MOFAObject)
}



#' Data quality control
#'
#' @param file_tmp DNA methylation or chromatin accessibility matrix
#' @param flag Choose to perform quality control by column or row
#'
#' @return Data after quality control
#' @export
#'
#' @examples
QC_meth_unmeth <- function(file_tmp,flag="col"){
  file_data <- file_tmp

  file_meth_name <- grep("\\.meth$", colnames(file_data), value = TRUE)
  file_unmeth_name <- grep("\\.UNmeth$", colnames(file_data), value = TRUE)

  file_meth_data <- file_data[,file_meth_name]
  file_unmeth_data <- file_data[,file_unmeth_name]

  file_meth_unmeth <- file_meth_data+file_unmeth_data
  rownames(file_meth_unmeth) <- rownames(file_data)
  file_name <- sub("\\.meth$","",file_meth_name)
  colnames(file_meth_unmeth) <- paste0(file_name,".meth_UNmeth")

  if (flag=="row") {
    file_data_num_count <- data.frame()
    file_data_num_count <- rownames(file_meth_unmeth)
    file_data_num_count <- as.data.frame(file_data_num_count)
    colnames(file_data_num_count)[1] <- "chr"

    file_meth_data_sum <- apply(file_meth_data, MARGIN = 1, sum)
    file_data_num_count$chr_meth_UNmeth_SUM <- apply(file_meth_unmeth, MARGIN = 1, sum)
    file_data_num_count$chr_methlevel <- file_meth_data_sum/file_data_num_count$chr_meth_UNmeth_SUM
    file_data_num_count_sorted <- arrange(file_data_num_count, -chr_meth_UNmeth_SUM)
    return(file_data_num_count_sorted)
  }else if (flag=="col") {
    file_data_num_count <- data.frame()
    file_data_num_count <- colnames(file_meth_unmeth)
    file_data_num_count <- as.data.frame(file_data_num_count)
    colnames(file_data_num_count)[1] <- "sample"

    file_meth_data_sum <- apply(file_meth_data, MARGIN = 2, sum)
    file_data_num_count$sample_meth_UNmeth_SUM <- apply(file_meth_unmeth, MARGIN = 2, sum)
    file_data_num_count$sample_methlevel <- file_meth_data_sum/file_data_num_count$sample_meth_UNmeth_SUM
    file_data_num_count_sorted <- arrange(file_data_num_count, -sample_meth_UNmeth_SUM)
    return(file_data_num_count_sorted)
  }
}






#' Statistics on the number of meths and UNmeths at different stages
#'
#' @param file_tmp
#' @param string
#'
#' @return
#' @export
#'
#' @examples
Meth_colname_sum <- function(file_tmp,string){
  file_tmp_col <- grep(string, colnames(file_tmp))

  file_tmp_string <- file_tmp[, file_tmp_col]
  rownames(file_tmp_string) <- rownames(file_tmp)
  file_tmp_string_rowsum <- rowSums(file_tmp_string)
  file_tmp_string_rowsum <- as.data.frame(file_tmp_string_rowsum)
  colnames(file_tmp_string_rowsum)[1] <- string
  return(file_tmp_string_rowsum)
}



#' The number of methylation levels in the same stage was counted through the .cov file
#'
#' @param dir cov format data generated by bismark
#' @param level_string Different sample information or developmental stages
#'
#' @return Data information of the same sample group or the same developmental stage
#' @export
#'
#' @examples
Meth_level_stage_count <- function(dir,level_string){
  all_files <- list.files(dir,pattern = "\\.cov$", full.names = TRUE)

  level_string_files <- grep(level_string, all_files, value = TRUE)

  level_string_data_list <- list()

  pb <- txtProgressBar(min = 0, max = 100, style = 3)
  cat("\n正在读取文件...\n")
  i=0

  for (file in level_string_files) {
    i=i+1
    file_dir <- paste0(dir,file)
    cov_data <- read.table(file_dir, header = F, sep = "\t", stringsAsFactors = FALSE)
    colnames(cov_data) <- c("chr", "start", "end", "methlevel" ,"meth" , "UNmeth")
    level_string_data_list <- c(level_string_data_list, list(cov_data))
    setTxtProgressBar(pb, i / length(level_string_files) * 100)
  }
  close(pb)
  # return(level_string_data_list)

  level_string_num <- length(level_string_files)
  level_string_data <- level_string_data_list[[1]]

  pb <- txtProgressBar(min = 0, max = 100, style = 3)
  cat("\n正在处理数据...\n")

  for (i in 2:level_string_num){
    difference_df1 <- anti_join(level_string_data, level_string_data_list[[i]], by = c("chr", "start", "end"))
    difference_df2 <- anti_join(level_string_data_list[[i]], level_string_data, by = c("chr", "start", "end"))

    intersection <- inner_join(level_string_data, level_string_data_list[[i]], by = c("chr", "start", "end"))

    unique_data <- rbind(difference_df1,difference_df2)

    result_df <- intersection[,c(1,2,3,5,6,8,9)]
    result_df$methSum <- result_df$meth.x+result_df$meth.y
    result_df$UNmethSum <- result_df$UNmeth.x+result_df$UNmeth.y

    result_df_deal <- result_df[,c(1,2,3,8,9)]
    result_df_deal_tmp <- result_df_deal$methSum/(result_df_deal$methSum+result_df_deal$UNmethSum)*100
    result_df_deal <- cbind(result_df_deal[,1:3],result_df_deal_tmp,result_df_deal[,4:ncol(result_df_deal)])
    colnames(result_df_deal) <- c("chr", "start", "end", "meth_level" ,"meth" , "UNmeth")

    merge_data <- rbind(unique_data,result_df_deal)
    level_string_data <- merge_data

    setTxtProgressBar(pb, i / level_string_num * 100)
  }
  close(pb)
  return(level_string_data)
}


#' Cov files to text files
#'
#' @param file_dir cov format data generated by bismark
#' @param region_split_file Processed physical fragment data, chrXX, xxx, xxx format
#'
#' @return DNA methylation or chromatin accessibility matrix
#' @export
#'
#' @examples
Cov_to_text <- function(file_dir,region_split_file){
  cov_files <- file_dir

  cov <- sprintf("%s:%s-%s", region_split_file$chr, region_split_file$start, region_split_file$end)
  chr_split <- region_split_file

  pb <- txtProgressBar(min = 0, max = 100, style = 3)
  cat("\n正在处理数据...\n")
  i=0

  for (i in 1:length(cov_files)) {
    cov_file <- read.table(cov_files[i])
    colnames(cov_file) <- c("chr","start","end","methlevel","meth","UNmeth")

    cov_file_name <- basename(cov_files[i])
    cov_name <- sub("\\..*", "", cov_file_name)
    cov_name_meth <- paste0(cov_name,".meth")
    cov_name_UNmeth <- paste0(cov_name,".UNmeth")
    cov_name_methlevel <- paste0(cov_name,".methlevel")

    mat <- matrix(nrow=nrow(chr_split), ncol=3)
    df_meth <- as.data.frame(mat)
    colnames(df_meth) <- c(cov_name_meth,cov_name_UNmeth,cov_name_methlevel)

    df_meth[,1] <- 0
    df_meth[,2] <- 0


    for(j in 1:nrow(chr_split)){
      for (k in 1:nrow(cov_file)) {
        if (chr_split$chr[j] == cov_file$chr[k] &&
            cov_file$start[k] >= chr_split$start[j] &&
            cov_file$start[k] <= chr_split$end[j]) {
          df_meth[j,1]=cov_file$meth[k]+df_meth[j,1]
          df_meth[j,2]=cov_file$UNmeth[k]+df_meth[j,2]
        }
      }
    }
    df_meth[,3] <- df_meth[,1]/(df_meth[,1]+df_meth[,2])
    cov <- cbind(cov,df_meth)
    setTxtProgressBar(pb, i / length(cov_files) * 100)
  }
  colnames(cov)[1] <- "chr"
  close(pb)
  return(cov)
}



#' Cov files to methlevel text files
#'
#' @param cov_file cov file path
#' @param cov_file_data Site data
#' @param chr_tmp Chromosome range
#' @param region Chromosome physical fragments
#'
#' @return DNA methylation or chromatin accessibility matrix
#' @export
#'
#' @examples
Cov_to_methleveltext <- function(cov_file, cov_file_data, chr_tmp, region) {
  # cov_file <- merge_CpG[1]
  # cov_file_data <- fread(merge_CpG[1])
  # chr_tmp <- chr_data

  region_chr <- region %>% filter(chr == chr_tmp)
  region_chr_paste <- sprintf("%s:%s-%s", region_chr$chr, region_chr$start, region_chr$end)
  region_chr_paste <- as.data.frame(region_chr_paste)
  colnames(region_chr_paste) <- "chr"

  cov_data <- cov_file_data
  colnames(cov_data) <- c("chr", "start", "end", "methlevel", "meth", "UNmeth")

  cov_data_chr <- cov_data %>% filter(chr == chr_tmp)

  # 生成列名
  cov_file_name <- basename(cov_file)
  split_result <- strsplit(cov_file_name, "\\.merged")[[1]][1]
  col_names <- paste0(split_result, c(".site", ".methlevel"))

  df_meth <- data.frame(matrix(nrow = nrow(region_chr_paste), ncol = 2))
  rownames(df_meth) <- region_chr_paste$chr
  colnames(df_meth) <- col_names

  # 使用向量化操作代替嵌套循环
  for (j in 1:nrow(region_chr)) {
    # 过滤出当前 region_chr 相关的 cov_data_chr
    filtered_cov_data <- cov_data_chr[cov_data_chr$start >= region_chr$start[j] & cov_data_chr$start <= region_chr$end[j], ]

    # 计算 site_count 和 methlevel_sum
    site_count <- nrow(filtered_cov_data)
    methlevel_sum <- ifelse(site_count > 0, sum(filtered_cov_data$methlevel), NA)

    # 赋值到 df_meth
    df_meth[j, ] <- c(site_count, methlevel_sum)
  }
  return(df_meth)
}



#' Filtering of site information
#'
#' @param files Site information file path
#' @param save_path Save folder path
#'
#' @return nothing
#' @export
#'
#' @examples
Cov_QC <- function(files,save_path){
  # files <- CpG_files
  # save_path <- "database_2/cov/STAD_SRP295352/CpG_filter/"

  for (i in 1:length(files)) {
    name <- basename(files[i])
    result <- sub("\\.cov$", "", name)
    result2 <- paste0(result, ".csv")

    file_data <- read.table(files[i])
    colnames(file_data) <- c("chr","start","end","methlevel","meth","UNmeth")

    file_data_filter <- file_data %>%
      filter(methlevel>=70 | methlevel<=30)
    file_data_save_path <- paste0(save_path, result2)

    # write.csv(file_data_filter,file_data_save_path,row.names = F)
    cat(files[i],"Success\n")
    fwrite(file_data_filter,file_data_save_path,sep = ',')
  }
}


#' According to HG19, the circle chart is displayed
#'
#' @param cov_file cov format data generated by bismark
#'
#' @return Circle Chart
#' @export
#'
#' @examples
Cov_circle <- function(cov_file){
  cov_file_choose <- STAD_GpC_merge_no0_unique[,c(1,2,3,4)]
  cov_file_choose[,4] <- cov_file_choose[,4]/100
  colnames(cov_file_choose) <- c("chr","start","end","methlevel")

  circos.initializeWithIdeogram(species = "hg19",plotType = c("axis"))

  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.rect(xlim[1], 0, xlim[2], 1, col = rand_color(24))
    circos.text(mean(xlim), mean(ylim), chr, cex = 0.6, col = "white",
                facing = "inside", niceFacing = TRUE)
  }, track.height = 0.1, bg.border = NA)

  circos.genomicTrack(cov_file_choose,
                      panel.fun = function(region, value, ...) {
                        circos.genomicPoints(region, value, col = "orange", pch = 16, cex = 0.6)
                      })

  col_fun <- colorRamp2(c(0, 0.5, 1), c("#82B0D2", "white", "#FF8884"))
  circos.genomicHeatmap(
    cov_file_choose,
    col = col_fun,
    side = "inside",          # 热图位置（inside/outside）
    heatmap_height = 0.15      # 轨道高度比例
  )
}






